src/disco/FixedWidthField/Real/ENDF/src/write.hpp

```diff
@@ -10,22 +10,68 @@ write( Representation real, Iterator& it ){
   } else if ( isInvalid( real ) ){
     writeInvalid( it );
   } else {
-    std::array< char, w + 3 > buffer;
-    auto greedyExponent = requiresTwoDigitExponent( real );
-    const auto length = w + 2;
-    const auto decimalDigits = 6 - greedyExponent;
-    int consumedCharacters = snprintf( &buffer[0], length + 1, "%*.*E",
-                                       length, decimalDigits, real );
-    if ( consumedCharacters < 0 ){
-      throw std::exception();
+
+    // Minimal exponent size is 2 (one sign and one digit)
+    constexpr int MIN_EXPWIDTH = 2;
+    // Minimal exponent for which we can use notation
+    constexpr int MIN_EXPONENT = -2;
+
+    // Decompose real
+    double significand{ real };
+    int exponent{ 0 };
+    int expWidth{ MIN_EXPWIDTH };
+    if( real != 0.0 ){
+      exponent = std::floor( std::log10( std::abs( real ) ) );
+      if( 0 < exponent ){
+        significand /= std::pow( 10.0, exponent );
+        expWidth += std::floor( std::log10( std::abs( exponent ) ) );
+      } else if( 0 > exponent ){
+        significand *= std::pow( 10.0, -exponent );
+        expWidth += std::floor( std::log10( std::abs( exponent ) ) );
+      }
```

Comment:
> Since you're rewriting perfectly good code, why not just write:
> 
> ```
> if ( 0 == exponent ) {
>   significand /= std::pow( 10.0, exponent );
>   expWidth += std::floor( std::log10( std::abs( exponent ) ) );
> }
> ```
> 
> The only reason there was an if on the sign of the exponent is that the pow function I wrote to calculate base^n assumes n was unsigned. Since you want to use std::pow, the exponent can be negative.

I can't do that because if `0 == exponent` then `std::log10( 0 )` is negative infinity; things break.

I was intentionally letting the compiler cast to an `int` instead of doing it via `static_cast`. Should be the same regardless. In fact, I we don't need the `std::floor` in there if we are casting to an int.

Reply:
This is why I should not review code late in the evening and you should not respond to a review early in the morning. It should have actually been:
```
if ( 0 != exponent ) {
  significand /= std::pow( 10.0, exponent );
  expWidth += std::floor( std::log10( std::abs( exponent ) ) );
}
```
About removing std::floor: having it there shows intent since that is what I want it to do. You should never rely on implicit conversions/casts to do work that can/could potentially go wrong. It takes away meaning and may be difficult to understand later on.

Reply:
Good call. I couldn't sleep because it was too hot. Probably too hot to think clearly too.

This is now improved.